src/Core/LanguageService.fs

```diff
@@ -732,64 +735,51 @@ Consider:
-                        |> Array.tryFindIndex (fun a -> a = "--fx-version")
-                        |> Option.map (fun _ -> true)
-                        |> Option.defaultValue false
-
-                    let shouldApplyImplicitRollForward = not (hasUserFxVersion || hasUserRollForward)
 
-                    let args =
-                        [ if shouldApplyImplicitRollForward then
-                              yield! rollForwardArgs
-                          yield! userDotnetArgs ]
-
-                    let envVariables =
-                        [ if shouldApplyImplicitRollForward then
-                              yield! necessaryEnvVariables ]
-
-                    return args, envVariables, fsacPath
+                    let! sdkVersionAtRootPath = sdkVersion ()
+
+                    match sdkVersionAtRootPath with
+                    | Error e ->
+                        printfn $"Error finding dotnet version: {e}"
+                        return failwith "Error finding dotnet version, do you have dotnet installed and on the PATH?"
+                    | Ok sdkVersion ->
+                        printfn "Parsed SDK version at root path: %s" sdkVersion.raw
+                        let sdkTfm = tfmForSdkVersion sdkVersion
+                        printfn "Parsed SDK version to tfm: %s" sdkTfm
+                        let fsacTfm, fsacPath = fsacPathForTfm sdkTfm
+                        printfn "Parsed TFM to fsac path: %s" fsacPath
+
+                        let userDotnetArgs = "FSharp.fsac.dotnetArgs" |> Configuration.get [||]
+
+                        let hasUserRollForward =
+                            userDotnetArgs
+                            |> Array.tryFindIndex (fun a -> a = "--roll-forward")
+                            |> Option.map (fun _ -> true)
+                            |> Option.defaultValue false
+
+                        let hasUserFxVersion =
+                            userDotnetArgs
+                            |> Array.tryFindIndex (fun a -> a = "--fx-version")
+                            |> Option.map (fun _ -> true)
+                            |> Option.defaultValue false
+
+                        let shouldApplyImplicitRollForward =
+                            not (hasUserFxVersion || hasUserRollForward)
+                            && sdkTfm <> fsacTfm // if the SDK doesn't match one of our FSAC TFMs, then we're in compat mode
+
+                        let args =
+                            userDotnetArgs
+
+                        let envVariables =
+                            [ if shouldApplyImplicitRollForward then
+                                "DOTNET_ROLL_FORWARD", box "LatestMajor"
+                              match sdkVersion.prerelease with
+                              | null -> ()
+                              | pres when pres.Count > 0 ->
+                                  "DOTNET_ROLL_FORWARD_TO_PRERELEASE", box "true"
+                              | _ -> () ]
```

Comment:
if the sdk in use is a preview we must apply another flag to let the runtime float forward to those previews